gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/GradleCacheExplorer.java

```diff
@@ -0,0 +1,155 @@
+            Set<String> stringsSet = new HashSet<>();
+            File gradleCacheFolder = new File(GRADLE_CACHE_PATH);
+            File[] metadataFolders = gradleCacheFolder.listFiles((dir, name) -> name.startsWith("metadata-"));
+
+            if (metadataFolders != null) {
+                for (File metadataFolder : metadataFolders) {
+                    if (indicator.isCanceled()) {
+                        throw new RuntimeException("Operation cancelled");
+                    }
+                    File binFile = new File(metadataFolder, "resource-at-url.bin");
+                    if (binFile.exists()) {
+                        extractStringsFromBinFile(binFile).stream()
+                                .filter(this::isValidResourceUrl)
+                                .map(this::sanitiseUrl)
+                                .forEach(stringsSet::add);
+                    }
+                }
+            }
+            return stringsSet;
+        });
+    }
+
+    final boolean isValidResourceUrl(String url) {
+        return projectUrls.stream().anyMatch(url::startsWith) && (url.endsWith(".pom") || url.endsWith(".jar"));
+    }
+
+    final Set<String> extractStringsFromBinFile(File binFile) {
+        Set<String> result = new HashSet<>();
+        try (BufferedInputStream bis = new BufferedInputStream(new FileInputStream(binFile))) {
+            StringBuilder currentString = new StringBuilder();
+            int byteValue;
+
+            while ((byteValue = bis.read()) != -1) {
+                char charValue = (char) byteValue;
+                if (isPrintableChar(charValue)) {
+                    currentString.append(charValue);
+                } else {
+                    if (currentString.length() >= 4) {
+                        result.add(currentString.toString());
+                    }
+                    currentString.setLength(0);
+                }
+            }
+
+            if (currentString.length() >= 4) {
+                result.add(currentString.toString());
+            }
+        } catch (IOException e) {
+            log.error("Failed to extract strings from bin file", e);
+        }
+
+        return result;
+    }
+
+    final boolean isPrintableChar(char ch) {
+        return ch >= 32 && ch <= 126; // Printable ASCII range
+    }
+
+    final String sanitiseUrl(String url) {
+        String finalUrl = url;
+        for (String projectUrl : projectUrls) {
+            if (finalUrl.startsWith(projectUrl)) {
+                finalUrl = finalUrl.substring(projectUrl.length());
+                break;
+            }
+        }
+
+        int lastSlashIndex = finalUrl.lastIndexOf('/');
+        int secondLastSlashIndex = finalUrl.lastIndexOf('/', lastSlashIndex - 1);
+
+        if (secondLastSlashIndex != -1) {
+            finalUrl = finalUrl.substring(0, secondLastSlashIndex).replace('/', '.');
+            finalUrl = finalUrl.replaceFirst("\\.([^.]*)$", ":$1");
+        }
```

Comment:
similarly here, if we don't find a second last slash index, the string returned is garbage no?

Reply:
so I would do something similar here and return an `Optional.empty`